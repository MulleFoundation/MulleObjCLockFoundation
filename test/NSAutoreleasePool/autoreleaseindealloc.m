#ifndef __MULLE_OBJC__
# import <Foundation/Foundation.h>
# pragma message "this test does not work with Apple Foundation"
#else
# import <MulleObjC/MulleObjC.h>
#endif


@interface Foo : NSObject
{
   id  _other;
}
@end

static int   n_instances;

@implementation Foo

+ (id) new
{
   ++n_instances;
   printf( "+new\n");
   return( [super new]);
}


- (void) dealloc
{
   [_other autorelease];

   --n_instances;
   printf( "-dealloc\n");
   [super dealloc];
}


- (void) setOther:(id) other
{
   [_other autorelease];
   _other = [other retain];
}


+ (void) test
{
   Foo   *foo;
   Foo   *other;

   //
   // MulleObjC should create enclosing autorelease pool
   // automatically

   foo = [[Foo new] autorelease];
   other = [[Foo new] autorelease];
   [foo setOther:other];
}

@end



int  main()
{
   // no crash or leak and we iz happy
   @autoreleasepool
   {
      [Foo test];
      printf( "%d\n", n_instances);  // 2
   }
   printf( "%d\n", n_instances);  // 0
   return( 0);
}
