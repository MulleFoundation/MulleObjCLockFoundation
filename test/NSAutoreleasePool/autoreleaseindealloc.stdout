+new
+new
2
-dealloc
-dealloc
0
